llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -3155,25 +3165,151 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+  case dwarf::DW_FORM_strx2:
+  case dwarf::DW_FORM_addrx2: {
+    return 2;
+  }
+  case dwarf::DW_FORM_strx3: {
+    return 3;
+  }
+  case dwarf::DW_FORM_data4:
+  case dwarf::DW_FORM_ref4:
+  case dwarf::DW_FORM_ref_sup4:
+  case dwarf::DW_FORM_strx4:
+  case dwarf::DW_FORM_addrx4: {
+    return 4;
+  }
+  case dwarf::DW_FORM_ref_sig8:
+  case dwarf::DW_FORM_data8:
+  case dwarf::DW_FORM_ref8:
+  case dwarf::DW_FORM_ref_sup8: {
+    return 8;
+  }
+  case dwarf::DW_FORM_data16: {
+    return 16;
+  }
+
+  case dwarf::DW_FORM_strp:
+  case dwarf::DW_FORM_sec_offset:
+  case dwarf::DW_FORM_GNU_ref_alt:
+  case dwarf::DW_FORM_GNU_strp_alt:
+  case dwarf::DW_FORM_line_strp:
+  case dwarf::DW_FORM_strp_sup: {
+    auto FormSize = FP.getDwarfOffsetByteSize();
+    return FormSize;
+  }
+  case dwarf::DW_FORM_addrx3:
+  case dwarf::DW_FORM_lo_user: {
+    llvm_unreachable("usupported form");
+    break;
+  }
+  }
+}
+
+Error DIEVisitor::materializeAbbrevDIE(unsigned AbbrevIdx) {
+  constexpr auto AddrSize = 8;
+  constexpr auto FormParams =
+      dwarf::FormParams{4 /*Version*/, AddrSize, dwarf::DwarfFormat::DWARF32};
+
   AbbrevEntryReader AbbrevReader =
       getAbbrevEntryReader(AbbrevEntries, AbbrevIdx);
 
-  if (Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag())
-    StartTagCallback(*MaybeTag, AbbrevIdx);
-  else
+  Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag();
+  if (!MaybeTag)
     return MaybeTag.takeError();
 
   Expected<bool> MaybeHasChildren = AbbrevReader.readHasChildren();
   if (!MaybeHasChildren)
     return MaybeHasChildren.takeError();
 
-  if (auto E = visitDIEAttrs(AbbrevReader, DataReader, DIEData))
+  AbbrevDIETagAndChildren TagAndChildren = {*MaybeTag, *MaybeHasChildren};
+  SmallVector<AbbrevDIEContents> AbbrevVector;
+  while (true) {
+    AbbrevDIEContents Contents;
+    Expected<dwarf::Attribute> Attr = AbbrevReader.readAttr();
+    if (!Attr)
+      return Attr.takeError();
+    if (*Attr == getEndOfAttributesMarker())
+      break;
+
+    Expected<dwarf::Form> Form = AbbrevReader.readForm();
+    if (!Form)
+      return Form.takeError();
+    Contents.Attr = *Attr;
+    Contents.Form = *Form;
+    Contents.FormInDistinctData = doesntDedup(*Form, *Attr);
+    auto FormSize = getNonULEBFormSize(*Form, FormParams);
+    Contents.FormSize = FormSize;
+    AbbrevVector.push_back(Contents);
+  }
+  AbbrevDIECache.push_back(std::make_pair(TagAndChildren, AbbrevVector));
+  return Error::success();
+}
+
+Error DIEVisitor::visitDIERef(BinaryStreamReader &DataReader,
+                              unsigned AbbrevIdx, StringRef DIEData,
+                              ArrayRef<DIEDataRef> &DIEChildrenStack) {
+
+  AbbrevDIECache.reserve(AbbrevEntries.size());
+  for (unsigned I = 0; I < AbbrevEntries.size(); I++) {
+    if (Error E = materializeAbbrevDIE(encodeAbbrevIndex(I)))
+      return E;
+  }
+
+  auto AbbrevDIECacheVal =
+      AbbrevDIECache[decodeAbbrevIndexAsAbbrevSetIdx(AbbrevIdx)];
```

Comment:
we usually use the find() method because operator[] default-constructs the value and thus silently fails if there is no cache hit.

Reply:
This should be more evident once we disable the default constructor for that struct

Reply:
I agree with Felipe, that is a seperate issue and I will put up a separate patch for that :)

Reply:
I am confused, `AbbrevDIECache` is a `SmallVector`, not a `DenseMap`

Reply:
Ah nevermind.

Reply:
So how does this function work? It materializes all abbreviations (3299) every time an entry is needed? I must be missing something obvious...

Reply:
The issue has been fixed, that was an oversight which did work when this patch was on top of the patch that switched materialization to iterative from recursive, but now I have a constructor in DIEVisitor to do the AbbrevEntryCache creation instead